src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ReliableRequestSessionChannel.cs

```diff
@@ -0,0 +1,842 @@
+                shutdownHandle.Set();
+            }
+
+            if (request != null)
+                request.Complete();
+        }
+
+        private async Task TerminateSequenceAsync(TimeSpan timeout)
+        {
+            CreateTerminateRequestor();
+            Message terminateReply = await terminateRequestor.RequestAsync(timeout);
+
+            if (terminateReply != null)
+            {
+                ProcessCloseOrTerminateReply(false, terminateReply);
+            }
+        }
+
+        private void UnblockClose()
+        {
+            FaultPendingRequests();
+
+            if (connection != null)
+            {
+                connection.Fault(this);
+            }
+
+            if (shutdownHandle != null)
+            {
+                shutdownHandle.Fault(this);
+            }
+
+            ReliableRequestor tempRequestor = closeRequestor;
+            if (tempRequestor != null)
+            {
+                tempRequestor.Fault(this);
+            }
+
+            tempRequestor = terminateRequestor;
+            if (tempRequestor != null)
+            {
+                tempRequestor.Fault(this);
+            }
+        }
+
+        private Task WaitForShutdownAsync(TimeSpan timeout)
+        {
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+
+            if (settings.ReliableMessagingVersion == ReliableMessagingVersion.WSReliableMessagingFebruary2005)
+            {
+                return shutdownHandle.WaitAsync(timeoutHelper.RemainingTime());
+            }
+            else
+            {
+                isLastKnown = true;
+
+                // We already closed the connection so we know everything was acknowledged.
+                // Make sure the reply acknowledgement ranges are current.
+                return replyAckConsistencyGuard.CloseAsync(timeoutHelper.RemainingTime());
+            }
+        }
+
+        private interface IReliableRequest : IRequestBase
+        {
+            void Set(Message reply);
+            void Complete();
+        }
+
+        private class AsyncRequest : IReliableRequest, IAsyncRequest
+        {
+            private bool _aborted = false;
+            private bool _completed = false;
+            private TaskCompletionSource<object> _tcs;
+            private bool _faulted = false;
+            private TimeSpan _originalTimeout;
+            private Message _reply;
+            private ReliableRequestSessionChannel _parent;
+
+            public AsyncRequest(ReliableRequestSessionChannel parent)
+            {
+                _parent = parent;
+            }
+
+            private object ThisLock { get; } = new object();
+
+            public void Abort(RequestChannel channel)
+            {
+                lock (ThisLock)
+                {
+                    if (!_completed)
+                    {
+                        _aborted = true;
+                        _completed = true;
+                        _tcs?.TrySetResult(null);
```

Comment:
Here and below: as above, consider using SetResult instead of TrySetResult

Reply:
This is in the abort code path. Abort can be called by user code at any time, or called via a timer callback. Calling _tcs.TrySetResult isn't guarded by a lock everywhere. If ReceiveReplyAsync needed to wait, it completes the TCS outside of a lock so this can be set multiple times. We don't care about the result, it's only being used as a latch.